### PR TITLE
{Misc} Fix name 'xxx' is not defined error in __main__.py

### PR DESCRIPTION
**Pull request by bebound**
**Description**

In [these issues](https://github.com/Azure/azure-cli/issues?q=is%3Aissue+is%3Aopen+%22name+%27exit_code%27+is+not+defined%22), the error message `Auto upgrade failed. name 'exit_code' is not defined` is useless.

I think `handle_exception` should be able to handle all exceptions, but sometimes it fails (In the above issues, most of them are result from import error). And `knack` raise error in [invoke()](https://github.com/microsoft/knack/blob/11f2454b8e4d977797057231d0850e7935b26fa6/knack/cli.py#L245), and `exit_code` is not defined.
https://github.com/Azure/azure-cli/blob/359a85d804eeac3b7a06407594150faf122977d8/src/azure-cli-core/azure/cli/core/util.py#L56

So I set `exit_code` to None, which makes the logic robust and removes `Name 'exit_code' can be undefined` in Python linter.

I also fix the same error for `az_upgrade_run`, `upgrade_exit_code` and `invoke_finish_time`.

---

This checklist is used to make sure that common guidelines for a pull request are followed.

- [ ] The PR title and description has followed the guideline in [Submitting Pull Requests](https://github.com/Azure/azure-cli/tree/dev/doc/authoring_command_modules#submitting-pull-requests).

- [ ] I adhere to the [Command Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/command_guidelines.md).

- [ ] I adhere to the [Error Handling Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/error_handling_guidelines.md).

### CODE REVIEW

**Comment by yonzhan** [comment]
Misc

File: src/azure-cli/azure/cli/__main__.py [hunk 1]
```diff
@@ -61,15 +62,12 @@ def cli_main(cli, args):
...
 
 finally:
-    try:
-        # Log the invoke finish time
-        invoke_finish_time = timeit.default_timer()
-        logger.info("Command ran in %.3f seconds (init: %.3f, invoke: %.3f)",
-                    invoke_finish_time - start_time,
-                    init_finish_time - start_time,
-                    invoke_finish_time - init_finish_time)
-    except NameError:
```
Comment (bebound) [inline src/azure-cli/azure/cli/__main__.py:71]: `NameError` is introduced in https://github.com/adewaleo/azure-cli/blob/26a3ddf7330b2958ed0fa40cfb74cd8477013ff7/src/azure-cli/azure/cli/__main__.py#L58-L62.
I think `NameError` is not raised unless `telemetry.start()` fails.

Comment (jiasli) [reply src/azure-cli/azure/cli/__main__.py:71]: Indeed, `NameError` should never be **caught and handled** by us. Instead, it should be **avoided** in the code itself.

File: src/azure-cli/azure/cli/__main__.py [hunk 2]
```diff
@@ -91,6 +89,8 @@ def cli_main(cli, args):
                     update_all = az_cli.config.getboolean('auto-upgrade', 'all', True)
                     prompt = az_cli.config.getboolean('auto-upgrade', 'prompt', True)
                     cmd = ['az', 'upgrade', '--all', str(update_all)]
+                    az_upgrade_run = False
+                    upgrade_exit_code = None
```
Comment (jiasli) [inline src/azure-cli/azure/cli/__main__.py:93]: PyCharm actually already warns about the `NameError`:

image

**Comment by jiasli** [comment]
While this PR makes auto-upgrade more robust, we still need to figure out why `ModuleNotFoundError: No module named 'jmespath'` happens and makes `handle_exception` crash.

**Review by jiasli — APPROVED** [review]

**Comment by bebound** [comment]
Most of the errors result from unofficial package, except https://github.com/Azure/azure-cli/issues/23716, which installs from Homebrew.
I'll investigate it if more users report this.

**Pull request merged**